File: architecture-skills/design-patterns/examples/builder/builder.js
// Builder pattern in JavaScript.
//
// Note: an options object with destructuring is usually more idiomatic for
// simple cases. Use a builder for multi-stage construction with validation.

class HttpRequest {
  constructor({ method, url, headers, body, timeoutSeconds }) {
    this.method = method;
    this.url = url;
    this.headers = headers;
    this.body = body;
    this.timeoutSeconds = timeoutSeconds;
    Object.freeze(this.headers);
    Object.freeze(this); // immutable product
  }
}

class HttpRequestBuilder {
  #method;
  #url;
  #headers = {};
  #body = null;
  #timeoutSeconds = 30;

  constructor(method, url) {
    this.#method = method.toUpperCase();
    this.#url = url;
  }

  header(name, value) {
    this.#headers[name] = value;
    return this;
  }

  body(body) {
    this.#body = body;
    return this;
  }

  timeout(seconds) {
    this.#timeoutSeconds = seconds;
    return this;
  }

  build() {
    if (this.#body !== null && ["GET", "HEAD"].includes(this.#method)) {
      throw new Error(`${this.#method} requests cannot have a body`);
    }
    return new HttpRequest({
      method: this.#method,
      url: this.#url,
      headers: { ...this.#headers },
      body: this.#body,
      timeoutSeconds: this.#timeoutSeconds,
    });
  }
}

const request = new HttpRequestBuilder("POST", "https://api.example.com/users")
  .header("Content-Type", "application/json")
  .header("Authorization", "Bearer token123")
  .body(JSON.stringify({ name: "Ada" }))
  .timeout(5)
  .build();

console.log(request);

File: architecture-skills/design-patterns/examples/builder/builder.py
"""Builder pattern in Python.

Note: for simple cases, keyword arguments with defaults are more idiomatic.
Use a builder when construction is multi-stage or needs validation.
"""

from __future__ import annotations

from dataclasses import dataclass, field


@dataclass(frozen=True)
class HttpRequest:
    """Immutable product."""

    method: str
    url: str
    headers: dict[str, str] = field(default_factory=dict)
    body: str | None = None
    timeout_seconds: float = 30.0


class HttpRequestBuilder:
    def __init__(self, method: str, url: str) -> None:
        self._method = method.upper()
        self._url = url
        self._headers: dict[str, str] = {}
        self._body: str | None = None
        self._timeout = 30.0

    def header(self, name: str, value: str) -> HttpRequestBuilder:
        self._headers[name] = value
        return self

    def body(self, body: str) -> HttpRequestBuilder:
        self._body = body
        return self

    def timeout(self, seconds: float) -> HttpRequestBuilder:
        self._timeout = seconds
        return self

    def build(self) -> HttpRequest:
        if self._body is not None and self._method in ("GET", "HEAD"):
            raise ValueError(f"{self._method} requests cannot have a body")
        return HttpRequest(
            method=self._method,
            url=self._url,
            headers=dict(self._headers),
            body=self._body,
            timeout_seconds=self._timeout,
        )


if __name__ == "__main__":
    request = (
        HttpRequestBuilder("POST", "https://api.example.com/users")
        .header("Content-Type", "application/json")
        .header("Authorization", "Bearer token123")
        .body('{"name": "Ada"}')
        .timeout(5.0)
        .build()
    )
    print(request)
